casr/src/bin/casr-ubsan.rs

```diff
@@ -0,0 +1,303 @@
+};
+use log::{debug, info};
+use rayon::iter::{IntoParallelRefIterator, ParallelIterator};
+
+use std::collections::HashMap;
+use std::fs;
+use std::fs::OpenOptions;
+use std::io::Write;
+use std::path::{Path, PathBuf};
+use std::process::Command;
+use std::sync::{Arc, Mutex};
+
+fn gen_report(
+    input: &PathBuf,
+    argv: &[&str],
+    pre_report: &CrashReport,
+    reports: &Arc<Mutex<HashMap<PathBuf, Vec<CrashReport>>>>,
+    counter: &Arc<Mutex<usize>>,
+) -> Result<()> {
+    let mut argv = argv.to_owned();
+    let pos = argv.iter().position(|&val| val == "@@").unwrap();
+    let input_arg = &input.clone().into_os_string().into_string().unwrap();
+    argv[pos] = input_arg;
+    // Run program.
+    let mut cmd = Command::new(argv[0]);
+    if argv.len() > 1 {
+        cmd.args(&argv[1..]);
+    }
+    debug!("{:?}", cmd);
+    let result = cmd
+        .output()
+        .with_context(|| "Couldn't run target program")?;
+    // Get stderr
+    let stderr = String::from_utf8_lossy(&result.stderr);
+    let stderr_list: Vec<&str> = stderr.split('\n').collect();
+
+    // Extract ubsan responds
+    let ubsan_reports = ubsan::extract_ubsan_reports(&stderr_list)?;
+    for ubsan_report in &ubsan_reports {
+        // Create report
+        let mut report = pre_report.clone();
+        report.proc_cmdline = argv.join(" ");
+        report.ubsan_report = ubsan_report.clone();
+        if ubsan_report[1].contains(": note: ") {
+            report.stacktrace = Vec::from(&ubsan_report[2..ubsan_report.len() - 2]);
+        } else {
+            report.stacktrace = Vec::from(&ubsan_report[1..ubsan_report.len() - 2]);
+        }
+        if let Some(exception) = ubsan::UbsanException::parse_exception(
+            &(ubsan_report.first().unwrap().to_owned() + ubsan_report.last().unwrap()),
+        ) {
+            report.execution_class = exception;
+        }
+        if let Ok(crash_line) = AsanStacktrace::parse_stacktrace(&report.stacktrace)?.crash_line() {
+            report.crashline = crash_line.to_string();
+            if let CrashLine::Source(debug) = crash_line {
+                if let Some(sources) = CrashReport::sources(&debug) {
+                    report.source = sources;
+                }
+            }
+        }
+        // Add report to hashmap
+        if let Some(reports_by_input) = reports.lock().unwrap().get_mut(input) {
+            reports_by_input.push(report);
+        }
+    }
+    // Update counter
+    let mut counter = counter.lock().unwrap();
+    *counter += ubsan_reports.len();
+    Ok(())
+}
+
+fn save_report(
+    report: &CrashReport,
+    matches: &ArgMatches,
+    input: &Path,
+    crashline: &str,
+) -> Result<()> {
+    // Convert report to string.
+    let repstr = serde_json::to_string_pretty(&report).unwrap();
+
+    // Get input dir name
+    let dir_name = input.parent().unwrap().file_name().unwrap();
+    let input_name = input.file_name().unwrap();
+    // Get crashline file name and line num
```

Comment:
Crash line parsing should be in ubsan.rs

Reply:
In ubsan.rs we get CrashLine struct from ubsan message, but there we need to parse crashline from string (report.crashline: String) with another structure, so we have to write 2 different crashline parsers anyway, I guess

Reply:
No longer relevant